File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-com-5-4/EnvironmentalProduct.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class EnvironmentalProduct : ISerialized
    {
        public EnvironmentalProduct(XmlNode xmlNode)
        {
            XmlNode productTypeNode = xmlNode.SelectSingleNode("productType");
            
            if (productTypeNode != null)
            {
                if (productTypeNode.Attributes["href"] != null || productTypeNode.Attributes["id"] != null) 
                {
                    if (productTypeNode.Attributes["id"] != null) 
                    {
                        productTypeIDRef_ = productTypeNode.Attributes["id"].Value;
                        EnvironmentalProductTypeEnum ob = new EnvironmentalProductTypeEnum(productTypeNode);
                        IDManager.SetID(productTypeIDRef_, ob);
                    }
                    else if (productTypeNode.Attributes["href"] != null)
                    {
                        productTypeIDRef_ = productTypeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        productType_ = new EnvironmentalProductTypeEnum(productTypeNode);
                    }
                }
                else
                {
                    productType_ = new EnvironmentalProductTypeEnum(productTypeNode);
                }
            }
            
        
            XmlNode compliancePeriodNode = xmlNode.SelectSingleNode("compliancePeriod");
            
            if (compliancePeriodNode != null)
            {
                if (compliancePeriodNode.Attributes["href"] != null || compliancePeriodNode.Attributes["id"] != null) 
                {
                    if (compliancePeriodNode.Attributes["id"] != null) 
                    {
                        compliancePeriodIDRef_ = compliancePeriodNode.Attributes["id"].Value;
                        EnvironmentalProductComplaincePeriod ob = new EnvironmentalProductComplaincePeriod(compliancePeriodNode);
                        IDManager.SetID(compliancePeriodIDRef_, ob);
                    }
                    else if (compliancePeriodNode.Attributes["href"] != null)
                    {
                        compliancePeriodIDRef_ = compliancePeriodNode.Attributes["href"].Value;
                    }
                    else
                    {
                        compliancePeriod_ = new EnvironmentalProductComplaincePeriod(compliancePeriodNode);
                    }
                }
                else
                {
                    compliancePeriod_ = new EnvironmentalProductComplaincePeriod(compliancePeriodNode);
                }
            }
            
        
            XmlNodeList vintageNodeList = xmlNode.SelectNodes("vintage");
            
            if (vintageNodeList != null)
            {
                this.vintage_ = new List<XsdTypeGYear>();
                foreach (XmlNode item in vintageNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            vintageIDRef_ = item.Attributes["id"].Value;
                            vintage_.Add(new XsdTypeGYear(item));
                            IDManager.SetID(vintageIDRef_, vintage_[vintage_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            vintageIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        vintage_.Add(new XsdTypeGYear(item));
                        }
                    }
                    else
                    {
                        vintage_.Add(new XsdTypeGYear(item));
                    }
                }
            }
            
        
            XmlNode applicableLawNode = xmlNode.SelectSingleNode("applicableLaw");
            
            if (applicableLawNode != null)
            {
                if (applicableLawNode.Attributes["href"] != null || applicableLawNode.Attributes["id"] != null) 
                {
                    if (applicableLawNode.Attributes["id"] != null) 
                    {
                        applicableLawIDRef_ = applicableLawNode.Attributes["id"].Value;
                        EnvironmentalProductApplicableLaw ob = new EnvironmentalProductApplicableLaw(applicableLawNode);
                        IDManager.SetID(applicableLawIDRef_, ob);
                    }
                    else if (applicableLawNode.Attributes["href"] != null)
                    {
                        applicableLawIDRef_ = applicableLawNode.Attributes["href"].Value;
                    }
                    else
                    {
                        applicableLaw_ = new EnvironmentalProductApplicableLaw(applicableLawNode);
                    }
                }
                else
                {
                    applicableLaw_ = new EnvironmentalProductApplicableLaw(applicableLawNode);
                }
            }
            
        
            XmlNode trackingSystemNode = xmlNode.SelectSingleNode("trackingSystem");
            
            if (trackingSystemNode != null)
            {
                if (trackingSystemNode.Attributes["href"] != null || trackingSystemNode.Attributes["id"] != null) 
                {
                    if (trackingSystemNode.Attributes["id"] != null) 
                    {
                        trackingSystemIDRef_ = trackingSystemNode.Attributes["id"].Value;
                        EnvironmentalTrackingSystem ob = new EnvironmentalTrackingSystem(trackingSystemNode);
                        IDManager.SetID(trackingSystemIDRef_, ob);
                    }
                    else if (trackingSystemNode.Attributes["href"] != null)
                    {
                        trackingSystemIDRef_ = trackingSystemNode.Attributes["href"].Value;
                    }
                    else
                    {
                        trackingSystem_ = new EnvironmentalTrackingSystem(trackingSystemNode);
                    }
                }
                else
                {
                    trackingSystem_ = new EnvironmentalTrackingSystem(trackingSystemNode);
                }
            }
            
        
        }
        
    
        #region ProductType_
        private EnvironmentalProductTypeEnum productType_;
        public EnvironmentalProductTypeEnum ProductType_
        {
            get
            {
                if (this.productType_ != null)
                {
                    return this.productType_; 
                }
                else if (this.productTypeIDRef_ != null)
                {
                    productType_ = IDManager.getID(productTypeIDRef_) as EnvironmentalProductTypeEnum;
                    return this.productType_; 
                }
                else
                {
                      return this.productType_; 
                }
            }
            set
            {
                if (this.productType_ != value)
                {
                    this.productType_ = value;
                }
            }
        }
        #endregion
        
        public string productTypeIDRef_ { get; set; }
        #region CompliancePeriod_
        private EnvironmentalProductComplaincePeriod compliancePeriod_;
        public EnvironmentalProductComplaincePeriod CompliancePeriod_
        {
            get
            {
                if (this.compliancePeriod_ != null)
                {
                    return this.compliancePeriod_; 
                }
                else if (this.compliancePeriodIDRef_ != null)
                {
                    compliancePeriod_ = IDManager.getID(compliancePeriodIDRef_) as EnvironmentalProductComplaincePeriod;
                    return this.compliancePeriod_; 
                }
                else
                {
                      return this.compliancePeriod_; 
                }
            }
            set
            {
                if (this.compliancePeriod_ != value)
                {
                    this.compliancePeriod_ = value;
                }
            }
        }
        #endregion
        
        public string compliancePeriodIDRef_ { get; set; }
        #region Vintage_
        private List<XsdTypeGYear> vintage_;
        public List<XsdTypeGYear> Vintage_
        {
            get
            {
                if (this.vintage_ != null)
                {
                    return this.vintage_; 
                }
                else if (this.vintageIDRef_ != null)
                {
                    return this.vintage_; 
                }
                else
                {
                      return this.vintage_; 
                }
            }
            set
            {
                if (this.vintage_ != value)
                {
                    this.vintage_ = value;
                }
            }
        }
        #endregion
        
        public string vintageIDRef_ { get; set; }
        #region ApplicableLaw_
        private EnvironmentalProductApplicableLaw applicableLaw_;
        public EnvironmentalProductApplicableLaw ApplicableLaw_
        {
            get
            {
                if (this.applicableLaw_ != null)
                {
                    return this.applicableLaw_; 
                }
                else if (this.applicableLawIDRef_ != null)
                {
                    applicableLaw_ = IDManager.getID(applicableLawIDRef_) as EnvironmentalProductApplicableLaw;
                    return this.applicableLaw_; 
                }
                else
                {
                      return this.applicableLaw_; 
                }
            }
            set
            {
                if (this.applicableLaw_ != value)
                {
                    this.applicableLaw_ = value;
                }
            }
        }
        #endregion
        
        public string applicableLawIDRef_ { get; set; }
        #region TrackingSystem_
        private EnvironmentalTrackingSystem trackingSystem_;
        public EnvironmentalTrackingSystem TrackingSystem_
        {
            get
            {
                if (this.trackingSystem_ != null)
                {
                    return this.trackingSystem_; 
                }
                else if (this.trackingSystemIDRef_ != null)
                {
                    trackingSystem_ = IDManager.getID(trackingSystemIDRef_) as EnvironmentalTrackingSystem;
                    return this.trackingSystem_; 
                }
                else
                {
                      return this.trackingSystem_; 
                }
            }
            set
            {
                if (this.trackingSystem_ != value)
                {
                    this.trackingSystem_ = value;
                }
            }
        }
        #endregion
        
        public string trackingSystemIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-com-5-4/EnvironmentalProduct.cpp
// EnvironmentalProduct.cpp 
#include "EnvironmentalProduct.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

EnvironmentalProduct::EnvironmentalProduct(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //productTypeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* productTypeNode = xmlNode->FirstChildElement("productType");

   if(productTypeNode){productTypeIsNull_ = false;}
   else{productTypeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- productTypeNode , address : " << productTypeNode << std::endl;
   #endif
   if(productTypeNode)
   {
      if(productTypeNode->Attribute("href") || productTypeNode->Attribute("id"))
      {
          if(productTypeNode->Attribute("id"))
          {
             productTypeIDRef_ = productTypeNode->Attribute("id");
             productType_ = boost::shared_ptr<EnvironmentalProductTypeEnum>(new EnvironmentalProductTypeEnum(productTypeNode));
             productType_->setID(productTypeIDRef_);
             IDManager::instance().setID(productTypeIDRef_,productType_);
          }
          else if(productTypeNode->Attribute("href")) { productTypeIDRef_ = productTypeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { productType_ = boost::shared_ptr<EnvironmentalProductTypeEnum>(new EnvironmentalProductTypeEnum(productTypeNode));}
   }

   //compliancePeriodNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* compliancePeriodNode = xmlNode->FirstChildElement("compliancePeriod");

   if(compliancePeriodNode){compliancePeriodIsNull_ = false;}
   else{compliancePeriodIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- compliancePeriodNode , address : " << compliancePeriodNode << std::endl;
   #endif
   if(compliancePeriodNode)
   {
      if(compliancePeriodNode->Attribute("href") || compliancePeriodNode->Attribute("id"))
      {
          if(compliancePeriodNode->Attribute("id"))
          {
             compliancePeriodIDRef_ = compliancePeriodNode->Attribute("id");
             compliancePeriod_ = boost::shared_ptr<EnvironmentalProductComplaincePeriod>(new EnvironmentalProductComplaincePeriod(compliancePeriodNode));
             compliancePeriod_->setID(compliancePeriodIDRef_);
             IDManager::instance().setID(compliancePeriodIDRef_,compliancePeriod_);
          }
          else if(compliancePeriodNode->Attribute("href")) { compliancePeriodIDRef_ = compliancePeriodNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { compliancePeriod_ = boost::shared_ptr<EnvironmentalProductComplaincePeriod>(new EnvironmentalProductComplaincePeriod(compliancePeriodNode));}
   }

   //vintageNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* vintageNode = xmlNode->FirstChildElement("vintage");

   if(vintageNode){vintageIsNull_ = false;}
   else{vintageIsNull_ = true;}

   if(vintageNode)
   {
      for(vintageNode; vintageNode; vintageNode = vintageNode->NextSiblingElement("vintage")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- vintageNode , address : " << vintageNode << std::endl;
          #endif
          if(vintageNode->Attribute("href") || vintageNode->Attribute("id"))
          {
              if(vintageNode->Attribute("id"))
              {
                  vintageIDRef_ = vintageNode->Attribute("id");
                  vintage_.push_back(boost::shared_ptr<XsdTypeGYear>(new XsdTypeGYear(vintageNode)));
                  vintage_.back()->setID(vintageIDRef_);
                  IDManager::instance().setID(vintageIDRef_, vintage_.back());
              }
              else if(vintageNode->Attribute("href")) { vintageIDRef_ = vintageNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { vintage_.push_back(boost::shared_ptr<XsdTypeGYear>(new XsdTypeGYear(vintageNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- vintageNode , address : " << vintageNode << std::endl;
       #endif
   }

   //applicableLawNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* applicableLawNode = xmlNode->FirstChildElement("applicableLaw");

   if(applicableLawNode){applicableLawIsNull_ = false;}
   else{applicableLawIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- applicableLawNode , address : " << applicableLawNode << std::endl;
   #endif
   if(applicableLawNode)
   {
      if(applicableLawNode->Attribute("href") || applicableLawNode->Attribute("id"))
      {
          if(applicableLawNode->Attribute("id"))
          {
             applicableLawIDRef_ = applicableLawNode->Attribute("id");
             applicableLaw_ = boost::shared_ptr<EnvironmentalProductApplicableLaw>(new EnvironmentalProductApplicableLaw(applicableLawNode));
             applicableLaw_->setID(applicableLawIDRef_);
             IDManager::instance().setID(applicableLawIDRef_,applicableLaw_);
          }
          else if(applicableLawNode->Attribute("href")) { applicableLawIDRef_ = applicableLawNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { applicableLaw_ = boost::shared_ptr<EnvironmentalProductApplicableLaw>(new EnvironmentalProductApplicableLaw(applicableLawNode));}
   }

   //trackingSystemNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* trackingSystemNode = xmlNode->FirstChildElement("trackingSystem");

   if(trackingSystemNode){trackingSystemIsNull_ = false;}
   else{trackingSystemIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- trackingSystemNode , address : " << trackingSystemNode << std::endl;
   #endif
   if(trackingSystemNode)
   {
      if(trackingSystemNode->Attribute("href") || trackingSystemNode->Attribute("id"))
      {
          if(trackingSystemNode->Attribute("id"))
          {
             trackingSystemIDRef_ = trackingSystemNode->Attribute("id");
             trackingSystem_ = boost::shared_ptr<EnvironmentalTrackingSystem>(new EnvironmentalTrackingSystem(trackingSystemNode));
             trackingSystem_->setID(trackingSystemIDRef_);
             IDManager::instance().setID(trackingSystemIDRef_,trackingSystem_);
          }
          else if(trackingSystemNode->Attribute("href")) { trackingSystemIDRef_ = trackingSystemNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { trackingSystem_ = boost::shared_ptr<EnvironmentalTrackingSystem>(new EnvironmentalTrackingSystem(trackingSystemNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<EnvironmentalProductTypeEnum> EnvironmentalProduct::getProductType()
{
   if(!this->productTypeIsNull_){
        if(productTypeIDRef_ != ""){
             return boost::shared_static_cast<EnvironmentalProductTypeEnum>(IDManager::instance().getID(productTypeIDRef_));
        }else{
             return this->productType_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<EnvironmentalProductTypeEnum>();
   }
}
boost::shared_ptr<EnvironmentalProductComplaincePeriod> EnvironmentalProduct::getCompliancePeriod()
{
   if(!this->compliancePeriodIsNull_){
        if(compliancePeriodIDRef_ != ""){
             return boost::shared_static_cast<EnvironmentalProductComplaincePeriod>(IDManager::instance().getID(compliancePeriodIDRef_));
        }else{
             return this->compliancePeriod_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<EnvironmentalProductComplaincePeriod>();
   }
}
std::vector<boost::shared_ptr<XsdTypeGYear>> EnvironmentalProduct::getVintage()
{
   if(!this->vintageIsNull_){
        if(vintageIDRef_ != ""){
             return this->vintage_;
        }else{
             return this->vintage_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<XsdTypeGYear>>();
   }
}
boost::shared_ptr<EnvironmentalProductApplicableLaw> EnvironmentalProduct::getApplicableLaw()
{
   if(!this->applicableLawIsNull_){
        if(applicableLawIDRef_ != ""){
             return boost::shared_static_cast<EnvironmentalProductApplicableLaw>(IDManager::instance().getID(applicableLawIDRef_));
        }else{
             return this->applicableLaw_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<EnvironmentalProductApplicableLaw>();
   }
}
boost::shared_ptr<EnvironmentalTrackingSystem> EnvironmentalProduct::getTrackingSystem()
{
   if(!this->trackingSystemIsNull_){
        if(trackingSystemIDRef_ != ""){
             return boost::shared_static_cast<EnvironmentalTrackingSystem>(IDManager::instance().getID(trackingSystemIDRef_));
        }else{
             return this->trackingSystem_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<EnvironmentalTrackingSystem>();
   }
}
}

